pkg/node/manager.go

```diff
@@ -64,12 +67,195 @@ func deleteNodeCIDR(ip *net.IPNet) {
 	}
 }
 
-func updateNodeCIDR(ip *net.IPNet, host net.IP) {
+func ipFamily(ip net.IP) int {
+	if ip.To4() == nil {
+		return netlink.FAMILY_V6
+	}
+
+	return netlink.FAMILY_V4
+}
+
+// findAddress finds a particular IP address on a link
+func findAddress(link netlink.Link, ip net.IP) *netlink.Addr {
+	addrs, err := netlink.AddrList(link, ipFamily(ip))
+	if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			logfields.IPAddr:    ip,
+			logfields.Interface: link.Attrs().Name,
+		}).Warn("Listing of addresses failed")
+
+		// return address not found on error
+		return nil
+	}
+
+	for _, a := range addrs {
+		if ip.Equal(a.IP) {
+			return &a
+		}
+	}
+
+	return nil
+}
+
+func findRoute(link netlink.Link, route *netlink.Route) *netlink.Route {
+	routes, err := netlink.RouteList(link, ipFamily(route.Dst.IP))
+	if err != nil {
+		return nil
+	}
+
+	for _, r := range routes {
+		aMaskLen, aMaskBits := r.Dst.Mask.Size()
+		bMaskLen, bMaskBits := route.Dst.Mask.Size()
+		if r.LinkIndex == route.LinkIndex && r.Scope == route.Scope &&
+			aMaskLen == bMaskLen && aMaskBits == bMaskBits &&
+			r.Dst.IP.Equal(route.Dst.IP) && r.Gw.Equal(route.Gw) {
+			return &r
+		}
+	}
+
+	return nil
+}
+
+func replaceNexthopRoute(link netlink.Link, routerIP, routerNet *net.IPNet) error {
+	// Add the Cilium router IP as address to the "cilium_host" if not
+	// already assigned
+	if routerIP != nil && findAddress(link, routerIP.IP) == nil {
+		addr := &netlink.Addr{IPNet: routerIP}
+		if err := netlink.AddrAdd(link, addr); err != nil {
+			return fmt.Errorf("unable to add nexthop address \"%s\": %q", routerIP, err)
+		}
+
+		log.WithFields(log.Fields{
+			logfields.IPAddr:    routerIP,
+			logfields.Interface: link.Attrs().Name,
+		}).Info("Added Cilium router IP address")
+	}
+
+	// This is the L2 route which makes the Cilium router IP available behind
+	// the "cilium_host" interface. All other routes will use this router IP
+	// as nexthop.
+	route := &netlink.Route{
+		LinkIndex: link.Attrs().Index,
+		Dst:       routerNet,
+		Scope:     netlink.SCOPE_LINK,
+	}
+	scopedLog := log.WithField(fieldRoute, route)
+
+	if findRoute(link, route) == nil {
+		if err := netlink.RouteDel(&netlink.Route{
+			LinkIndex: link.Attrs().Index,
+			Dst:       routerNet,
+		}); err != nil {
+			scopedLog.WithError(err).Debug("No old L2 nexthop route exists")
+		}
+
+		if err := netlink.RouteAdd(route); err != nil {
+			scopedLog.WithError(err).Error("Unable to add L2 nexthop route")
+			return fmt.Errorf("unable to add L2 nexthop route")
+		}
+
+		scopedLog.Info("Added L2 nexthop route")
+	}
+
+	return nil
+}
+
+func replaceNodeRoute(ip *net.IPNet) {
```

Comment:
Can you add a description to that function of what it does?

Reply:
fixed